Implement the "info" command

info: Print connection information.

- [x] Look up tokio docs for getting client information
- [x] Find a way to pass client information into the service
- [x] Send the formatted client information as the response
- [x] Format like the original cliserver

```
   Compiling cliserver_rust v0.1.0 (file:///home/nitrogen/devel/learning_rust/cliserver_rust)      
error: `io` does not live long enough                                                              
   --> src/main.rs:120:25                                                                          
    |                                                                                              
120 |         codec.io = Some(io.get_ref());                                                       
    |                         ^^ does not live long enough                                         
...                                                                                                
123 |     }                                                                                        
    |     - borrowed value only lives until here                                                   
    |                                                                                              
    = note: borrowed value must be valid for the static lifetime...                                
                                                                                                   
error[E0382]: use of moved value: `io`                                                             
   --> src/main.rs:120:25                                                                          
    |                                                                                              
118 |         let q: Self::BindTransport = Ok(io);                                                 
    |                                         -- value moved here                                  
119 |                                                                                              
120 |         codec.io = Some(io.get_ref());                                                       
    |                         ^^ value used here after move                                        
    |                                                                                              
    = note: move occurs because `io` has type `tokio_core::io::Framed<T, CliCodec<T>>`, which does 
not implement the `Copy` trait                                                                     
                                                                                                   
error: aborting due to 2 previous errors                                                           
                                                                                                   
error: Could not compile `cliserver_rust`.                                                         
```

Some of the earlier problems I was having with "cannot infer type for '_'" errors may have been caused by a stray semicolon at the end of the implicit return line.  These compiler errors could use some improvement.  Here are the related errors:

```
error[E0282]: unable to infer enough type information about `_`
   --> src/main.rs:118:9
    |
118 |         Ok(io.framed(codec) as Self::Transport);
    |         ^^ cannot infer type for `_`
    |
    = note: type annotations or generic parameter binding required
```

and

```
error[E0269]: not all control paths return a value                
   --> src/main.rs:113:5                                          
    |                                                             
113 |     fn bind_transport(&self, io: T) -> Self::BindTransport {
    |     ^                                                       
```

A new error while trying to treat `TcpStream` differently from other `Io` objects:

```
error[E0531]: unresolved tuple struct/variant `tokio_core::net::TcpStream`
   --> src/main.rs:115:13
    |
115 |             tokio_core::net::TcpStream(tcp) => {
    |             ^^^^^^^^^^^^^^^^^^^^^^^^^^
```

--- a/src/main.rs
+++ b/src/main.rs
@@ -8,6 +8,7 @@ use std::str;
 use std::collections::BTreeMap;
 use tokio_core::io::{Codec, EasyBuf};
 use tokio_core::io::{Io, Framed};
+use tokio_core::net::TcpStream;
 use tokio_proto::pipeline::ServerProto;
 use tokio_proto::TcpServer;
 use tokio_service::Service;
@@ -103,26 +104,16 @@ impl Codec for CliCodec {
 
 pub struct CliProto;
 
-impl<T: Io + 'static + std::fmt::Debug> ServerProto<T> for CliProto {
+impl ServerProto<TcpStream> for CliProto {
     type Request = (String, String, Option<String>);
     type Response = String;
 
-    type Transport = Framed<T, CliCodec>;
+    type Transport = Framed<TcpStream, CliCodec>;
     type BindTransport = Result<Self::Transport, io::Error>;
 
-    fn bind_transport(&self, io: T) -> Self::BindTransport {
-        let info = match io {
-            tokio_core::net::TcpStream(tcp) => {
-                let addr = tcp.peer_addr().unwrap();
-                format!(
-                    "Client address: {}\nClient port: {}\n",
-                    addr.ip, addr.port
-                    )
-            },
-            _ => {
-                format!("{:?}", &io)
-            }
-        };
+    fn bind_transport(&self, io: TcpStream) -> Self::BindTransport {
+        let addr = io.peer_addr().unwrap();
+        let info = format!("Client address: {}\nClient port: {}", addr.ip(), addr.port());
         Ok(io.framed(CliCodec::new(info)))
     }
 }

--- a/src/main.rs
+++ b/src/main.rs
@@ -13,15 +13,24 @@ use tokio_proto::TcpServer;
 use tokio_service::Service;
 use futures::{future, Future, BoxFuture};
 
-#[derive(Default)]
-pub struct CliCodec {
+pub struct CliCodec<T: Io + std::fmt::Debug + 'static> {
+    /// Underlying I/O object used by this codec
+    io: Option<&'static T>,
+
     /// Offset within the incoming EasyBuf at which newline search last ended
     search_offset: usize,
 }
 
-impl Codec for CliCodec {
-    /// Input type is a tuple with the command name and an optional argument string
-    type In = (String, Option<String>);
+impl<T: Io + std::fmt::Debug + 'static> CliCodec<T> {
+    /// Initializes with None for io; the IO must be set later
+    fn new() -> CliCodec<T> {
+        CliCodec { io: None, search_offset: 0 }
+    }
+}
+
+impl<T: Io + std::fmt::Debug + 'static> Codec for CliCodec<T> {
+    /// Input type is a tuple with the I/O, the command name, and an optional argument string
+    type In = (&'static T, String, Option<String>);
 
     /// Response type is just a string
     type Out = String;
@@ -55,9 +64,13 @@ impl Codec for CliCodec {
                         let request = match s.find(' ') {
                             Some(i) => {
                                 let (cmd, args) = s.split_at(i);
-                                (cmd.trim().to_string(), Some(args.trim().to_string()))
+                                (
+                                    self.io.unwrap(),
+                                    cmd.trim().to_string(),
+                                    Some(args.trim().to_string())
+                                )
                             },
-                            None => (s.to_string(), None)
+                            None => (self.io.unwrap(), s.to_string(), None)
                         };
 
                         Ok(Some(request))
@@ -90,39 +103,51 @@ impl Codec for CliCodec {
 
 pub struct CliProto;
 
-impl<T: Io + 'static> ServerProto<T> for CliProto {
-    type Request = (String, Option<String>);
+impl<T: Io + std::fmt::Debug + 'static> ServerProto<T> for CliProto {
+    type Request = (&'static T, String, Option<String>);
     type Response = String;
 
-    type Transport = Framed<T, CliCodec>;
+    type Transport = Framed<T, CliCodec<T>>;
     type BindTransport = Result<Self::Transport, io::Error>;
 
     fn bind_transport(&self, io: T) -> Self::BindTransport {
-        Ok(io.framed(CliCodec::default()))
+        println!("IO received by bind_transport: {:?}", &io);
+
+        let mut codec = CliCodec::new();
+        let io = io.framed(codec);
+        let q: Self::BindTransport = Ok(io);
+
+        codec.io = Some(io.get_ref());
+
+        q
     }
 }
 
 
-trait CliCommand {
+trait CliCommand<T: Io + std::fmt::Debug + 'static> {
     fn name(&self) -> &str;
     fn description(&self) -> &str;
-    fn call(&self, args: Option<String>) -> String;
+    fn call(&self, io: &T, args: Option<String>) -> String;
 }
 
 
 #[derive(Default)]
-pub struct CliServer<'a> {
-    commands: BTreeMap<&'a str, &'a CliCommand>,
+pub struct CliServer<'a, T: Io + std::fmt::Debug + 'static> {
+    commands: BTreeMap<&'a str, &'a CliCommand<T>>,
 }
 
-impl<'a> CliServer<'a> {
-    fn add_command(&mut self, cmd: &'a CliCommand) {
+impl<'a, T: Io + std::fmt::Debug + 'static> CliServer<'a, T> {
+    fn new() -> CliServer<'a, T> {
+        CliServer { commands: BTreeMap::new() }
+    }
+
+    fn add_command(&mut self, cmd: &'a CliCommand<T>) {
         self.commands.insert(cmd.name(), cmd);
     }
 }
 
-impl<'a> Service for CliServer<'a> {
-    type Request = (String, Option<String>);
+impl<'a, T: Io + std::fmt::Debug + 'static> Service for CliServer<'a, T> {
+    type Request = (&'static T, String, Option<String>);
     type Response = String;
 
     type Error = io::Error;
@@ -130,15 +155,17 @@ impl<'a> Service for CliServer<'a> {
     type Future = BoxFuture<Self::Response, Self::Error>;
 
     fn call(&self, req: Self::Request) -> Self::Future {
-        match self.commands.get(&req.0[..]) {
+        let (io, cmdname, args) = req;
+
+        match self.commands.get(&cmdname[..]) {
             Some(cmd) => {
-                println!("Calling command {}", req.0);
+                println!("Calling command {} for IO {:?}", cmdname, io);
                 // TODO: have commands return futures?
-                future::ok(cmd.call(req.1)).boxed()
+                future::ok(cmd.call(io, args)).boxed()
             },
             None => {
-                println!("No match found for command {}", req.0);
-                future::ok(format!("Unknown command: {}", req.0)).boxed()
+                println!("No match found for command {}", cmdname);
+                future::ok(format!("Unknown command: {}", cmdname)).boxed()
             }
         }
     }
@@ -147,7 +174,7 @@ impl<'a> Service for CliServer<'a> {
 // TODO: Consider registering commands by passing two strings and a closure
 struct EchoCommand;
 
-impl CliCommand for EchoCommand {
+impl<T: Io + std::fmt::Debug + 'static> CliCommand<T> for EchoCommand {
     fn name(&self) -> &str {
         "echo"
     }
@@ -156,7 +183,7 @@ impl CliCommand for EchoCommand {
         "Prints the command line."
     }
 
-    fn call(&self, args: Option<String>) -> String {
+    fn call(&self, io: &T, args: Option<String>) -> String {
         match args {
             Some(s) => s,
             None => "".to_string()
@@ -172,7 +199,7 @@ fn main() {
 
     println!("Serving on {}", addr);
     server.serve(|| {
-        let mut cli = CliServer::default();
+        let mut cli = CliServer::new();
         cli.add_command(&ECHO);
         Ok(cli)
     });

--- a/src/main.rs
+++ b/src/main.rs
@@ -13,24 +13,24 @@ use tokio_proto::TcpServer;
 use tokio_service::Service;
 use futures::{future, Future, BoxFuture};
 
-pub struct CliCodec<T: Io + std::fmt::Debug + 'static> {
-    /// Underlying I/O object used by this codec
-    io: Option<&'static T>,
+pub struct CliCodec {
+    /// Description of the I/O object for the info command
+    info: String,
 
     /// Offset within the incoming EasyBuf at which newline search last ended
     search_offset: usize,
 }
 
-impl<T: Io + std::fmt::Debug + 'static> CliCodec<T> {
-    /// Initializes with None for io; the IO must be set later
-    fn new() -> CliCodec<T> {
-        CliCodec { io: None, search_offset: 0 }
+impl CliCodec {
+    /// Initializes with the given connection info
+    fn new(io_info: String) -> CliCodec {
+        CliCodec { info: io_info, search_offset: 0 }
     }
 }
 
-impl<T: Io + std::fmt::Debug + 'static> Codec for CliCodec<T> {
-    /// Input type is a tuple with the I/O, the command name, and an optional argument string
-    type In = (&'static T, String, Option<String>);
+impl Codec for CliCodec {
+    /// Input type is a tuple with the I/O info, the command name, and an optional argument string
+    type In = (String, String, Option<String>);
 
     /// Response type is just a string
     type Out = String;
@@ -65,12 +65,12 @@ impl<T: Io + std::fmt::Debug + 'static> Codec for CliCodec<T> {
                             Some(i) => {
                                 let (cmd, args) = s.split_at(i);
                                 (
-                                    self.io.unwrap(),
+                                    self.info.clone(),
                                     cmd.trim().to_string(),
                                     Some(args.trim().to_string())
                                 )
                             },
-                            None => (self.io.unwrap(), s.to_string(), None)
+                            None => (self.info.clone(), s.to_string(), None)
                         };
 
                         Ok(Some(request))
@@ -103,51 +103,47 @@ impl<T: Io + std::fmt::Debug + 'static> Codec for CliCodec<T> {
 
 pub struct CliProto;
 
-impl<T: Io + std::fmt::Debug + 'static> ServerProto<T> for CliProto {
-    type Request = (&'static T, String, Option<String>);
+impl<T: Io + 'static + std::fmt::Debug> ServerProto<T> for CliProto {
+    type Request = (String, String, Option<String>);
     type Response = String;
 
-    type Transport = Framed<T, CliCodec<T>>;
+    type Transport = Framed<T, CliCodec>;
     type BindTransport = Result<Self::Transport, io::Error>;
 
     fn bind_transport(&self, io: T) -> Self::BindTransport {
         println!("IO received by bind_transport: {:?}", &io);
 
-        let mut codec = CliCodec::new();
-        let io = io.framed(codec);
-        let q: Self::BindTransport = Ok(io);
-
-        codec.io = Some(io.get_ref());
-
-        q
+        // TODO: Mimic cliserver output format for TCP connections
+        let codec = CliCodec::new(format!("{:?}", &io));
+        Ok(io.framed(codec))
     }
 }
 
 
-trait CliCommand<T: Io + std::fmt::Debug + 'static> {
+trait CliCommand {
     fn name(&self) -> &str;
     fn description(&self) -> &str;
-    fn call(&self, io: &T, args: Option<String>) -> String;
+    fn call(&self, info: String, args: Option<String>) -> String;
 }
 
 
 #[derive(Default)]
-pub struct CliServer<'a, T: Io + std::fmt::Debug + 'static> {
-    commands: BTreeMap<&'a str, &'a CliCommand<T>>,
+pub struct CliServer<'a> {
+    commands: BTreeMap<&'a str, &'a CliCommand>,
 }
 
-impl<'a, T: Io + std::fmt::Debug + 'static> CliServer<'a, T> {
-    fn new() -> CliServer<'a, T> {
+impl<'a> CliServer<'a> {
+    fn new() -> CliServer<'a> {
         CliServer { commands: BTreeMap::new() }
     }
 
-    fn add_command(&mut self, cmd: &'a CliCommand<T>) {
+    fn add_command(&mut self, cmd: &'a CliCommand) {
         self.commands.insert(cmd.name(), cmd);
     }
 }
 
-impl<'a, T: Io + std::fmt::Debug + 'static> Service for CliServer<'a, T> {
-    type Request = (&'static T, String, Option<String>);
+impl<'a> Service for CliServer<'a> {
+    type Request = (String, String, Option<String>);
     type Response = String;
 
     type Error = io::Error;
@@ -155,13 +151,13 @@ impl<'a, T: Io + std::fmt::Debug + 'static> Service for CliServer<'a, T> {
     type Future = BoxFuture<Self::Response, Self::Error>;
 
     fn call(&self, req: Self::Request) -> Self::Future {
-        let (io, cmdname, args) = req;
+        let (info, cmdname, args) = req;
 
         match self.commands.get(&cmdname[..]) {
             Some(cmd) => {
-                println!("Calling command {} for IO {:?}", cmdname, io);
+                println!("Calling command {} for IO info {:?}", cmdname, info);
                 // TODO: have commands return futures?
-                future::ok(cmd.call(io, args)).boxed()
+                future::ok(cmd.call(info, args)).boxed()
             },
             None => {
                 println!("No match found for command {}", cmdname);
@@ -174,7 +170,7 @@ impl<'a, T: Io + std::fmt::Debug + 'static> Service for CliServer<'a, T> {
 // TODO: Consider registering commands by passing two strings and a closure
 struct EchoCommand;
 
-impl<T: Io + std::fmt::Debug + 'static> CliCommand<T> for EchoCommand {
+impl CliCommand for EchoCommand {
     fn name(&self) -> &str {
         "echo"
     }
@@ -183,7 +179,7 @@ impl<T: Io + std::fmt::Debug + 'static> CliCommand<T> for EchoCommand {
         "Prints the command line."
     }
 
-    fn call(&self, io: &T, args: Option<String>) -> String {
+    fn call(&self, _info: String, args: Option<String>) -> String {
         match args {
             Some(s) => s,
             None => "".to_string()

--- a/src/main.rs
+++ b/src/main.rs
@@ -111,11 +111,19 @@ impl<T: Io + 'static + std::fmt::Debug> ServerProto<T> for CliProto {
     type BindTransport = Result<Self::Transport, io::Error>;
 
     fn bind_transport(&self, io: T) -> Self::BindTransport {
-        println!("IO received by bind_transport: {:?}", &io);
-
-        // TODO: Mimic cliserver output format for TCP connections
-        let codec = CliCodec::new(format!("{:?}", &io));
-        Ok(io.framed(codec))
+        let info = match io {
+            tokio_core::net::TcpStream(tcp) => {
+                let addr = tcp.peer_addr().unwrap();
+                format!(
+                    "Client address: {}\nClient port: {}\n",
+                    addr.ip, addr.port
+                    )
+            },
+            _ => {
+                format!("{:?}", &io)
+            }
+        };
+        Ok(io.framed(CliCodec::new(info)))
     }
 }
 

--- a/src/main.rs
+++ b/src/main.rs
@@ -169,7 +169,6 @@ impl<'a> Service for CliServer<'a> {
 
 // TODO: Consider registering commands by passing two strings and a closure
 struct EchoCommand;
-
 impl CliCommand for EchoCommand {
     fn name(&self) -> &str {
         "echo"
@@ -187,7 +186,18 @@ impl CliCommand for EchoCommand {
     }
 }
 
+struct InfoCommand;
+impl CliCommand for InfoCommand {
+    fn name(&self) -> &str { "info" }
+    fn description(&self) -> &str { "Prints connection information." }
+
+    fn call(&self, info: String, _args: Option<String>) -> String {
+        info
+    }
+}
+
 static ECHO: EchoCommand = EchoCommand;
+static INFO: InfoCommand = InfoCommand;
 
 fn main() {
     let addr = "0.0.0.0:14311".parse().unwrap();
@@ -197,6 +207,7 @@ fn main() {
     server.serve(|| {
         let mut cli = CliServer::new();
         cli.add_command(&ECHO);
+        cli.add_command(&INFO);
         Ok(cli)
     });
 }

